File: Breadth_First_Search/2146.K_Highest_Ranked_Items_Within_a_Price_Range/Python/2146.K_Highest_Ranked_Items_Within_a_Price_Range.py
class Solution(object):
    def highestRankedKItems(self, grid, pricing, start, k):
        """
        :type grid: List[List[int]]
        :type pricing: List[int]
        :type start: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        #============================================#
        # Breath-first search based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(grid), len(grid[0])

        ##### Record array #####
        record_arry = [[0 for _ in range(len_cols)] for _ in range(len_rows)]

        ##### Record queue, heap/priority-queue #####
        record_queue, record_pq = [start], []

        heapify(record_pq)

        ##### Record distance #####
        record_dist = 0

        ##### Result array #####
        res_arry = []


        ############################################################################
        #Breath-first search based loop traversal with recorded heap/priority-queue
        ##### Step 1: Record informations with heap/priority-queue #####
        while (record_queue):
            queue_size = len(record_queue) #Queue-size

            for _ in range(queue_size):
                [record_row_idx, record_col_idx] = record_queue.pop(0) #Keep updating/overwriting

                ##### Check if the current position exceed boundary or not #####
                if (((record_row_idx < 0) or (record_row_idx >= len_rows)) or ((record_col_idx < 0) or (record_col_idx >= len_cols))):
                    continue

                ##### Check if the current position is visited or not #####
                if ((record_arry[record_row_idx])[record_col_idx]):
                    continue

                (record_arry[record_row_idx])[record_col_idx] = 1 #Keep updating/overwriting

                record_val = (grid[record_row_idx])[record_col_idx] #Keep updating/overwriting

                ##### Check if the current position is valid or not #####
                if (not record_val):
                    continue

                ##### Check if the current price is in ranges or not #####
                if ((record_val >= pricing[0]) and (record_val <= pricing[1])):
                    heappush(record_pq, [record_dist, record_val, record_row_idx, record_col_idx]) #Keep updating/recording
                else:
                    pass

                record_queue.append([record_row_idx, (record_col_idx - 1)]) #Keep updating/recording
                record_queue.append([(record_row_idx - 1), record_col_idx]) #Keep updating/recording
                record_queue.append([record_row_idx, (record_col_idx + 1)]) #Keep updating/recording
                record_queue.append([(record_row_idx + 1), record_col_idx]) #Keep updating/recording

            record_dist += 1 #Keep updating/accumulating

        ##### Step 2: Looped-traversal with recorded heap/priority-queue #####
        record_times = (k if (k < len(record_pq)) else len(record_pq)) #Record times 

        for record_idx in range(record_times):
            res_arry.append([(record_pq[0])[2], (record_pq[0])[3]]) #Keep updating/recording

            heappop(record_pq) #Keep updating/popped

        return res_arry

File: Breadth_First_Search/2146.K_Highest_Ranked_Items_Within_a_Price_Range/C++/2146.K_Highest_Ranked_Items_Within_a_Price_Range.cpp
class Solution {
public:
    vector<vector<int>> highestRankedKItems(vector<vector<int>> & grid, vector<int> & pricing, vector<int> & start, int k) 
    {
        //============================================//
        // Input type:                                //
        //   - vector<vector<int>> ref. (grid vector) //
        //   - vector<int> ref. (pricing vector)      //
        //   - vector<int> ref. (start vector)        //
        //   - int (k-th order)                       //
        // Return type:                               //
        //   - vector<vector<int>> (result vector)    //
        //============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = grid.size(), len_cols = (grid[0]).size();

        ///// Record vector /////
        vector<vector<int>> record_vect(len_rows, vector<int>(len_cols, 0));

        ///// Record queue /////
        queue<pair<int, int>> record_queue;

        ///// Record heap, priority-queue /////
        priority_queue<vector<int>> record_pq;

        ///// Record pair /////
        pair<int, int> record_pair;

        ///// Record row, column /////
        int record_row, record_col;

        ///// Record value, distance /////
        int record_val, record_dist = 0;

        ///// Result vector /////
        vector<vector<int>> res_vect;


        /*Breath-first search based loop traversal with recorded heap, priority-queue*/
        record_queue.push(make_pair(start[0], start[1])); //Update, Record

        ///// Step 1: Record informations with heap, priority-queue /////
        while (! record_queue.empty()) //Whole
        {
            for (int queue_idx = record_queue.size(); (queue_idx > 0); (queue_idx--)) //Queue-size
            {
                record_pair = record_queue.front(); record_queue.pop(); //Keep updating, overwriting

                record_row = record_pair.first; record_col = record_pair.second; //Keep updating, overwriting

                ///// Check if the current edge boundary is valid or not /////
                if (((record_row < 0) || (record_row >= len_rows)) || ((record_col < 0) || (record_col >= len_cols))) { continue; }

                ///// Check if the current position is valid or not /////
                if (! (grid[record_row])[record_col]) { continue; }

                ///// Check if the current position is visited or not /////
                if ((record_vect[record_row])[record_col]) { continue; }

                (record_vect[record_row])[record_col] = 1; //Keep updating, overwriting

                record_val = (grid[record_row])[record_col]; //Keep updating, overwriting

                ///// Check if the current value is in range or not /////
                if ((record_val >= pricing[0]) && (record_val <= pricing[1]))
                { 
                    record_pq.push({((-1) * record_dist), ((-1) * record_val), ((-1) * record_row), ((-1) * record_col)}); //Keep updating, recording
                }
                else { ; }

                record_queue.push(make_pair(record_row, (record_col - 1))); record_queue.push(make_pair((record_row - 1), record_col)); //Keep updating, recording
                record_queue.push(make_pair(record_row, (record_col + 1))); record_queue.push(make_pair((record_row + 1), record_col)); //Keep updating, recording

            } //Queue-size

            (record_dist++); //Keep updating, accumulating
        }

        ///// Step 2: Looped-traversal with recorded heap, priority-queue /////
        for (int record_idx = (k < record_pq.size()) ? k : record_pq.size(); (record_idx > 0); (record_idx--)) //Order-size
        {
            res_vect.push_back({((record_pq.top())[2] * (-1)), ((record_pq.top())[3] * (-1))}); //Keep updating, recording

            record_pq.pop(); //Keep updating, popped

        } //Order-size

        return res_vect;
    }
};
